### DISCUSSION TITLE
Router.on.events not firing

### DISCUSSION THREAD

**Discussion by clarencejychan** [vercel/next.js#11650] [Help]
I currently have a custom server in a test app 
```
// import App from 'next/app'

import { useEffect } from "react"
import Router from 'next/router'

function MyApp({ Component, pageProps }) {
    useEffect(() => {
        const handleRouteChange = url => {
          console.log('App is changing to: ', url)
        }
      
        Router.events.on('routeChangeStart', handleRouteChange)
        return () => {
          Router.events.off('routeChangeStart', handleRouteChange)
        }
    }, [])


    return <Component {...pageProps} />
  }
  
  // Only uncomment this method if you have blocking data requirements for
  // every single page in your application. This disables the ability to
  // perform automatic static optimization, causing every page in your app to
  // be server-side rendered.
  //
  // MyApp.getInitialProps = async (appContext) => {
  //   // calls page's `getInitialProps` and fills `appProps.pageProps`
  //   const appProps = await App.getInitialProps(appContext);
  //
  //   return { ...appProps }
  // }
  
  export default MyApp
```

which is nearly identical to the example that was given in the documents. When I go from "/" to "/test" on the browser, which is another page I've created in my test app, I can never see the console.log fire which seems to indicate that the events are never firing for whatever reason. 

I'm not sure if I"m doing something wrong or if this is a bug

**Comment by clarencejychan** [comment]
it seems that if i do a hard reload so if enter localhost:3000/ in the browser, the event never fires. If I enter localhost:3000/test it also never fires.

It seems that it will only fire on Route.push calls or <link> calls

Is there anyway to "intercept" the routes so that I can to redirect users on certain conditions on certain routes if they enter the route on the browser?

**Reply by borekb** [reply]
> it seems that if i do a hard reload so if enter localhost:3000/ in the browser, the event never fires. If I enter localhost:3000/test it also never fires.

That is expected, `'routeChangeStart'` happens when you navigate in the browser and is _not_ triggered during SSR.

**Reply by antkougblenou** [reply]
are there any of these events that are triggered during SSR ? and if not what can be used instead to get information about route change ?

**Reply by Magimart** [reply]
try out this, i hope i understood your problems well enough.
this will fire each time route changes after the component is mounted
and when the route changes due to component unmounting then the state is set to false
enabling firing on and firing out

U may want want to pass a var in the parameter and attaching it to the event.On() method.
Have a loook!
function MyApp({ Component, pageProps, router }) {
  const [onMount, setOnMount] = useState(true);

  useEffect(() => {
    const routeChange = () => {
      onMount && setOnMount(false);
    };

    router.events.on("startOnRouteChange", routeChange);

   
    return () => {
      router.events.off("startOnRouteChange", routeChange);
    };
  }, []);
  
  therefore onMount will return true on each mount

**Comment by prakhar-parashar-au7** [comment]
Heyy, I'm using next.js, this code works fine when I use it in [path].js , and when route is being changed from '/something' to '/somethingelse', 
but it doesn't work when I put it in index.js, and when I'm going from '/' to '/somethingElse'. Is there any reason why it doesn't work in index.js?

**Comment by dklear** [comment]
I was following the [documentation](https://nextjs.org/docs/api-reference/next/router#usage-6) for listening for all route changes on `_app.` The problem was that `handleRouteChange` does not get invoked unless you include `url` and `{ shallow }` parameters, which we drop.

**Comment by joaojbs199** [comment]
In my case it doesn't work because I added "target" to the <a> tag

**Comment by kmgdevelopment** [comment]
I'm also having this problem. Does it matter if the event listener is added in a sub-component vs the top-level app component? This is the code I currently have in a component imported into a dynamic `pages/[slug].tsx` page:

```
import { useRouter } from 'next/router';
import { useEffect } from "react";

export default function MyComponent() {
    const router = useRouter();
   
    useEffect( () => {
        const handleRouteChange = () => {
            console.log('NextJS route change event');
        }
        router.events.on('routeChangeComplete', handleRouteChange);

        return router.events.off('routeChangeComplete', handleRouteChange);
    },[router]);

    return (
        <div>Hello I am a component</div>
    )
}
```

This doesn't fire if I navigate from a subpage to the homepage, or if I go from one subpage to another subpage. As far as I can tell this is the exact code recommended in the docs so I'm stumped.

**Reply by kmgdevelopment** [reply]
My problem was that I was using `routerChangeComplete` instead of `routerChangeStart`. 

I got it to work with `routerChangeComplete` but only if I removed the cleanup function. I assume the reason is that the component unmounts _before_ `routerChangeComplete` fires, so the listener is removed before the event fires. I haven't tested it but my guess is `routerChangeComplete` is only meant to be used on the top-level App component since it never unmounts.

So for my specific problem using `routerChangeStart` gets me what I need and the cleanup function works.
